packages/core/src/storage/namespace.ts

```diff
@@ -0,0 +1,396 @@
+ */
+function getOriginalStruct(structCanonicalName: string, origContractDef: ContractDefinition) {
+  for (const node of origContractDef.nodes) {
+    if (isNodeType('StructDefinition', node)) {
+      if (node.canonicalName === structCanonicalName) {
+        return node;
+      }
+    }
+  }
+  throw new Error(
+    `Could not find original source location for namespace struct with name ${structCanonicalName} from contract ${origContractDef.name}`,
+  );
+}
+
+/**
+ * Gets the original source location for the given struct canonical name and struct member label.
+ */
+function getOriginalMemberSrc(structCanonicalName: string, memberLabel: string, origContractDef: ContractDefinition) {
+  const node = getOriginalStruct(structCanonicalName, origContractDef);
+  if (node !== undefined) {
+    const typeMembers = getTypeMembers(node, { src: true });
+    assert(typeMembers !== undefined);
+
+    for (const member of typeMembers) {
+      if (typeof member !== 'string') {
+        if (member.label === memberLabel) {
+          return member.src;
+        }
+      }
+    }
+  }
+}
+
+/**
+ * From the given layout types, gets the struct member matching the given struct canonical name and struct member label.
+ */
+function getStructMemberFromLayoutTypes(
+  types: Record<string, TypeItem<string>>,
+  structCanonicalName: string,
+  memberLabel: string,
+) {
+  const structType = findTypeWithLabel(types, `struct ${structCanonicalName}`);
+  const structMembers = structType?.members;
+  if (structMembers !== undefined) {
+    for (const structMember of structMembers) {
+      assert(typeof structMember !== 'string');
+      if (structMember.label === memberLabel) {
+        return structMember;
+      }
+    }
+  }
+  return undefined;
+}
+
+/**
+ * From the given layout types, gets the type matching the given type label.
+ */
+function findTypeWithLabel(types: Record<string, TypeItem>, label: string) {
+  for (const type of Object.values(types)) {
+    if (type.label === label) {
+      return type;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Makes a modified copy of the solc input to add state variables in each contract for namespaced struct definitions,
+ * so that the compiler will generate their types in the storage layout.
+ *
+ * This deletes all functions for efficiency, since they are not needed for storage layout.
+ * We also need to delete modifiers and immutable variables to avoid compilation errors due to deleted
+ * functions and constructors.
+ *
+ * Also sets the outputSelection to only include storageLayout and ast, since the other outputs are not needed.
+ *
+ * @param input The original solc input.
+ * @param output The original solc output.
+ * @returns The modified solc input with storage layout that includes namespaced type information.
+ */
+export function makeNamespacedInputCopy(input: SolcInput, output: SolcOutput): SolcInput {
+  const modifiedInput: SolcInput = JSON.parse(JSON.stringify(input));
```

Comment:
`JSON.parse(JSON.stringify(...))` is a quick hack that works, but for this use case I'm pretty sure it will be particularly inefficient. The intuition is that there are a lot of strings in this object (all of the Solidity source code) that stringify will copy into a new string before they are parsed again. If we used some procedure that copied the object in memory, the strings can just be copied by reference because JavaScript strings are immutable, so it should be a lot cheaper.

The native way to do a deep copy is `structuredClone`. It's not available in Node 16, which we still support, but we can use a polyfill: https://github.com/ungap/structured-clone

Reply:
Changed to copy parts of the solc input by reference, and construct the resulting solc input object using those references and the generated portions.